### ISSUE TITLE
Re-evaluate TestSubscriber API

### ISSUE THREAD

**Issue by smaldini** [reactor/reactor-core#135]
Currently the TestSubscriber works with post subscribe assertions. I collected during SpringOne conference some feedbacks from CloudFoundry Java client and Spring teams to reasonably question the current way of testing. It might mean that an immediate step is to not release is as part of the GA until we figure out what would be efficient or not (e.g. an expectation centric approach then subscribe).

This issue is a place to elaborate on that and we can take that discussion offline if required.

/cc @rstoyanchev @sdeleuze @akarnokd @nebhale @poutsma @sbrannen

**Comment by akarnokd** [comment]
`TestSubscriber` is an established way of testing operators and flows (now [more convenient](https://github.com/ReactiveX/RxJava/blob/2.x/src/test/java/io/reactivex/internal/operators/flowable/FlowableConcatMapEagerTest.java) in RxJava 2 with the `.test()` operator).

I'm curious what alternatives do you have in mind?

**Comment by poutsma** [comment]
Hi David,

While I do agree that `TestSubscriber` is valuable for the scenarios you described, we have found - in both Spring Reactive as well as our CloundFoundry Java client - that perhaps there is room for improvement, or at least an alternative approach. For instance, the before mentioned CloudFoundry project have had success with their own [tailored variant](https://github.com/cloudfoundry/cf-java-client/blob/2ff05ce58c49665fab29b4b10781910dfd57f6c2/cloudfoundry-util/src/main/java/org/cloudfoundry/util/test/TestSubscriber.java), which is based on setting up signal expectation upfront, as opposed to asserting them after the fact.

So one of the possibilities we want to investigate is to see if we can come up with more polished version of that, conceptually similar to mocking frameworks like EasyMock and others. In pseudo-code, something like this:

```java
MockSubscriber<String> mockSubscriber = MockSubscriber
   .expectValue("foo")
   .expectValue("bar")
  .expectComplete(); // or expectError(Throwable)

Flux<String> flux = Flux.just("foo", "bar");
flux.subscribe(mockSubscriber);
mockSubscriber.verify();
```

Given the fact that Reactor 3.0.GA has a time-driven release schedule, and keeping in mind Josh Bloch's [mantra](http://www.artima.com/intv/blochP.html) "when in doubt, leave it out", we are currently planning to make the TestSubscriber available through other means (through a separate artifact, or just as source code through Github), until we have investigated this post 3.0. The end result could very well be that we reintroduce `TestSubscriber`, but at this point we are simply not sure yet. So we leave it out 😃.

I hope this clears things up, and I am looking forward to reading your thoughts on this.

**Comment by JakeWharton** [comment]
I too find `TestSubscriber` to rarely be what I want.

Specifying a all of the values every time I need to assert one is annoying, and doesn't scale beyond the simple numbers that we see in RxJava's tests. We have our own subscriber which "consumes" a value from the seen list when you assert on it.

Another thing I often implement is one backed by a blocking queue so that asynchronous sources can be tested easily. This allows my tests to read like repeated chunks of "perform an action", "wait for these events", "assert the events were what I expected".

I've never found it a problem to just build my own test subscribers for these cases, and I would be sad to have it be absent from this library or RxJava 2.

**Comment by smaldini** [comment]
Removed from now as we roll up a new TestSubscriber for the upcoming versions.

**Comment by daniellavoie** [comment]
So... `TestSubscriber` has been removed from RC1 to RC2. What's the alternative ? All the tests we built are based on the 'TestSubscriber`.

**Comment by smaldini** [comment]
Workaround : Copy in your test the one here https://github.com/reactor/reactor-core/blob/master/src/test/java/reactor/test/TestSubscriber.java

Short-term : A new test module (segregated from core) might include it along with a WIP ScriptedSubscriber and TestScheduler.

**Comment by daniellavoie** [comment]
I'm buying ! I'll copy the private class in my code base. Thanks !

However, I am a little disturbed to witness breaking API changes between release candidates. Doesn't feel like a usual thing. I understand reactor is still under heavy development but I am concerned about adoption if API keeps changing. What is your point of view on this matter ?

**Comment by smaldini** [comment]
We knew that decision will be affecting even our early adopters and tbh it wasn't without pain we decided to pull off especially that we ourselves use it a lot. But we've seen early use of reactor and it was our last chance to call for such change before GA and 3.0.2 is a supported version (3.0.0 had a niche API signature issue thats why we waited 3.0.2 for the GA).

To mitigate that decision a few things came to mind : 
- You can still copy the class, its not like we removed an operator or changed the behavior of Mono.
- Better to have a support that internally is used (we've seen CF java client creating their own TestSubscriber for instance)
- There is a core design problem with such TestSubscriber that could lead unprepared user to confusion, especially concurrent assertions with flowing signals (you subscribed before and assert after).
- A few companies would be just happy to not have test code in production and moving it eventually as it was to a different module scope (test) felt not a bad idea.

So we did that under this very scope, we can't afford to do that for every API and Rx has enshrined a few already. Also Spring can't afford volatility neither. Similar to that move we might just move RxJava adapters into an extra addon, so that's the kind of breaking change at most you could expect from 3.0.x. 
For 3.1 there might be some statistic and tracking improvement in the basket at least.
For 4.0, not much is on the table apart from changing the interfaces we depend from RS jar to Java9 Flow. But this is very early to have any tangible stuff to tell you about at this stage.

So TL;DR, if we did it that's because we had a window, otherwise we are trying to build some trust for us (Spring) and beyond :)

**Comment by daniellavoie** [comment]
Thanks for taking the time to explain the context !

**Comment by smaldini** [comment]
Update you can start using back TestSubscriber from
io.projectreactor.addons:reactor-test:0.1.0.BUILD-SNAPSHOT. Versioning will be amended and released along with reactor-core:3.0.3.RELEASE. The project is hosted under https://github.com/reactor/reactor-addons/.

**Comment by smaldini** [comment]
We are also moving the PR https://github.com/reactor/reactor-core/pull/201 wrt an expectation-based testing subscriber.
